mitosheet/src/jupyter/notebook/pluginUtils.tsx

```diff
@@ -0,0 +1,255 @@
+import MitoAPI from "../api";
+import { containsGeneratedCodeOfAnalysis, containsMitosheetCallWithAnyAnalysisToReplay, containsMitosheetCallWithSpecificAnalysisToReplay, getArgsFromMitosheetCallCode, getCodeString, isMitosheetCallCode } from "../../utils/code";
+
+type CellType = any;
+
+export function getCellAtIndex(index: number): CellType | undefined {
+    return (window as any).Jupyter?.notebook?.get_cell(index);
+}
+
+export function getCellText(cell: CellType | undefined): string {
+    return cell?.get_text() || '';
+}
+
+
+/* 
+    Returns True if the passed cell is empty.
+    Returns False if the passed cells is either not empty or undefined 
+*/
+export function isEmptyCell(cell: CellType | undefined): boolean {
+    if (cell === undefined) {
+        return false;
+    }
+    const currentCode = getCellText(cell);
+    return currentCode.trim() === '';
+}
+
+/**
+ * Returns the cell that has the mitosheet.sheet(analysis_to_replay={analysisName}) in it,
+ * or undefined if no such cell exists
+ */
+export function getCellCallingMitoshetWithAnalysis(analysisName: string): [CellType, number] | undefined  {
+    const cells: CellType[] = (window as any).Jupyter?.notebook?.get_cells();
+
+    if (cells === undefined) {
+        return undefined;
+    }
+
+    let cellIndex = 0;
+    for (const cell of cells) {
+        if (containsMitosheetCallWithSpecificAnalysisToReplay(getCellText(cell), analysisName)) {
+            return [cell, cellIndex];
+        }
+
+        cellIndex++;
+    }
+
+    return undefined;
+}
+
+
+/**
+ * A function that returns the [cell, index] pair of the mitosheet.sheet() call that contains
+ * the analysis name. 
+ * 
+ * If no mitosheet.sheet() call contains this analysis name, then we assume it hasen't been 
+ * written yet, and take our best guess at which sheet this is.
+ * 
+ * Returns undefined if it can find no good guess for a calling mitosheet cell.
+ */
+export function getMostLikelyMitosheetCallingCell(analysisName: string | undefined): [CellType, number] | undefined {
+
+    // First, we check if this analysis name is in a mitosheet call, in which case things are easy
+    if (analysisName) {
+        const mitosheetCallCellAndIndex = getCellCallingMitoshetWithAnalysis(analysisName);
+        if (mitosheetCallCellAndIndex !== undefined) {
+            return mitosheetCallCellAndIndex;
+        }
+    }
+
+    const cells = (window as any).Jupyter?.notebook?.get_cells();
+
+    if (cells == undefined) {
+        return;
+    }
+
+    const activeCell = (window as any).Jupyter?.notebook?.get_cell((window as any).Jupyter?.notebook?.get_anchor_index());
```

Comment:
With this new architecture, having to do the code robustness PR that you recently did would be a bit frustrating, because you'd pretty much have to figure out the logic on lab, then copy it all over for notebooks. 

A different approach we could take is creating a lower level api that takes as a param the type of jupyter. 

ie: Here, we would use a function called `getActiveCell` that would return the active cell reguardless of if it is jupyter lab or jupyter notebook. 

For future compatability, this would reduce the branching if we, for example, do a vscode integration. 

On the other hand, who knows that that integration will look like and what the optimal architecture will be